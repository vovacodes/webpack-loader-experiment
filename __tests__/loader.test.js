const path = require('path');
const { execSync } = require("child_process");
const fs = require('fs-extra');
const webpack = require('webpack');
const memoryfs = require('memory-fs');

const tempDirPath = path.join(__dirname, '../tmp');

describe("loader", () => {
  afterAll(() => {
    fs.removeSync(tempDirPath);
  });

  it("should inject packages json data and update it when it changes", (done) => {
    const tempProjectDirPath = prepareFixtureProject('project1');
    const compiler = getCompiler(tempProjectDirPath);
    let buildCount = 0;

    const watching = compiler.watch({ aggregateTimeout: 0 }, (err, stats) => {
      if (err) return done(err);
      if (stats.hasErrors()) return done(new Error(stats.toJson().errors));

      buildCount++;
      const output = stats.toJson().modules[0].source;

      expect(output).toMatchSnapshot();

      if (buildCount === 1) {
        fs.writeFileSync(
          path.join(tempProjectDirPath, "node_modules", "nanoid", "package.json"),
          '{ "name": "modified" }'
        );
      }

      if (buildCount === 2) {
        watching.close(() => {
          done();
        });
      }
    });
  });

  it("should update packages data when package.json changes", (done) => {
    const tempProjectDirPath = prepareFixtureProject('project1');
    const compiler = getCompiler(tempProjectDirPath);
    let buildCount = 0;

    const watching = compiler.watch({ aggregateTimeout: 0 }, (err, stats) => {
      if (err) return done(err);
      if (stats.hasErrors()) return done(new Error(stats.toJson().errors));

      buildCount++;
      const output = stats.toJson().modules[0].source;

      expect(output).toMatchSnapshot();

      if (buildCount === 1) {
        fs.writeFileSync(
          path.join(tempProjectDirPath, "package.json"),
          `
            {
              "name": "project1",
              "version": "1.0.0",
              "main": "index.js",
              "license": "MIT",
              "dependencies": {
                "lodash.isarraylike": "^4.2.0"
              }
            }
          `
        );
      }

      if (buildCount === 2) {
        watching.close(() => {
          done();
        });
      }
    });
  });

  it("should not re-read the packages data if nothing changed", (done) => {
    const tempProjectDirPath = prepareFixtureProject('project1');
    const compiler = getCompiler(tempProjectDirPath);
    let buildCount = 0;

    const watching = compiler.watch({ aggregateTimeout: 0 }, (err, stats) => {
      if (err) return done(err);
      if (stats.hasErrors()) return done(new Error(stats.toJson().errors));

      buildCount++;
      const output = stats.toJson().modules[0].source;

      expect(output).toMatchSnapshot();

      if (buildCount === 1) {
        watching.invalidate()
      }

      if (buildCount === 2) {
        watching.close(() => {
          done();
        });
      }
    });
  });
});

function prepareFixtureProject(projectName) {
  const tempProjectDirPath = path.join(tempDirPath, projectName);
  const projectPath = path.join(__dirname, '../__fixtures__', projectName);

  fs.emptyDirSync(tempDirPath);
  fs.copySync(projectPath, tempProjectDirPath);
  execSync("yarn", { cwd: tempProjectDirPath });

  return tempProjectDirPath;
}

function getCompiler(projectDirPath) {
  const compiler = webpack({
    context: projectDirPath,
    entry: "../../__fixtures__/module.js",
    output: {
      path: "/",
      filename: 'bundle.js',
    },
    module: {
      rules: [{
        test: /module\.js$/,
        use: {
          loader: '../../lib/loader.js',
        }
      }]
    }
  });

  compiler.outputFileSystem = new memoryfs();

  return compiler;
}
