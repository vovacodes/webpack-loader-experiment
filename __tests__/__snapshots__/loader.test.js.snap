// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should inject packages json data and update it when it changes 1`] = `
"const packages = {};

packages[\\"lodash.isarraylike\\"] = {
  \\"name\\": \\"lodash.isarraylike\\",
  \\"version\\": \\"4.2.0\\",
  \\"description\\": \\"The lodash method \`_.isArrayLike\` exported as a module.\\",
  \\"homepage\\": \\"https://lodash.com/\\",
  \\"icon\\": \\"https://lodash.com/icon.svg\\",
  \\"license\\": \\"MIT\\",
  \\"keywords\\": \\"lodash-modularized, isarraylike\\",
  \\"author\\": \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
  \\"contributors\\": [
    \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
    \\"Blaine Bublitz <blaine.bublitz@gmail.com> (https://github.com/phated)\\",
    \\"Mathias Bynens <mathias@qiwi.be> (https://mathiasbynens.be/)\\"
  ],
  \\"repository\\": \\"lodash/lodash\\",
  \\"scripts\\": { \\"test\\": \\"echo \\\\\\"See https://travis-ci.org/lodash/lodash-cli for testing details.\\\\\\"\\" }
}

packages[\\"nanoid\\"] = {
  \\"name\\": \\"nanoid\\",
  \\"version\\": \\"2.0.3\\",
  \\"description\\": \\"A tiny (141 bytes), secure URL-friendly unique string ID generator\\",
  \\"keywords\\": [
    \\"uuid\\",
    \\"random\\",
    \\"id\\",
    \\"url\\"
  ],
  \\"author\\": \\"Andrey Sitnik <andrey@sitnik.ru>\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"ai/nanoid\\",
  \\"browser\\": {
    \\"./index.js\\": \\"./index.browser.js\\",
    \\"./random.js\\": \\"./random.browser.js\\",
    \\"./async/index.js\\": \\"./async/index.browser.js\\",
    \\"./async/random.js\\": \\"./async/random.browser.js\\"
  },
  \\"sideEffects\\": false
}


export default packages;
"
`;

exports[`loader should inject packages json data and update it when it changes 2`] = `
"const packages = {};

packages[\\"lodash.isarraylike\\"] = {
  \\"name\\": \\"lodash.isarraylike\\",
  \\"version\\": \\"4.2.0\\",
  \\"description\\": \\"The lodash method \`_.isArrayLike\` exported as a module.\\",
  \\"homepage\\": \\"https://lodash.com/\\",
  \\"icon\\": \\"https://lodash.com/icon.svg\\",
  \\"license\\": \\"MIT\\",
  \\"keywords\\": \\"lodash-modularized, isarraylike\\",
  \\"author\\": \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
  \\"contributors\\": [
    \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
    \\"Blaine Bublitz <blaine.bublitz@gmail.com> (https://github.com/phated)\\",
    \\"Mathias Bynens <mathias@qiwi.be> (https://mathiasbynens.be/)\\"
  ],
  \\"repository\\": \\"lodash/lodash\\",
  \\"scripts\\": { \\"test\\": \\"echo \\\\\\"See https://travis-ci.org/lodash/lodash-cli for testing details.\\\\\\"\\" }
}

packages[\\"nanoid\\"] = { \\"name\\": \\"modified\\" }

export default packages;
"
`;

exports[`loader should not re-read the packages data if nothing changed 1`] = `
"const packages = {};

packages[\\"lodash.isarraylike\\"] = {
  \\"name\\": \\"lodash.isarraylike\\",
  \\"version\\": \\"4.2.0\\",
  \\"description\\": \\"The lodash method \`_.isArrayLike\` exported as a module.\\",
  \\"homepage\\": \\"https://lodash.com/\\",
  \\"icon\\": \\"https://lodash.com/icon.svg\\",
  \\"license\\": \\"MIT\\",
  \\"keywords\\": \\"lodash-modularized, isarraylike\\",
  \\"author\\": \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
  \\"contributors\\": [
    \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
    \\"Blaine Bublitz <blaine.bublitz@gmail.com> (https://github.com/phated)\\",
    \\"Mathias Bynens <mathias@qiwi.be> (https://mathiasbynens.be/)\\"
  ],
  \\"repository\\": \\"lodash/lodash\\",
  \\"scripts\\": { \\"test\\": \\"echo \\\\\\"See https://travis-ci.org/lodash/lodash-cli for testing details.\\\\\\"\\" }
}

packages[\\"nanoid\\"] = {
  \\"name\\": \\"nanoid\\",
  \\"version\\": \\"2.0.3\\",
  \\"description\\": \\"A tiny (141 bytes), secure URL-friendly unique string ID generator\\",
  \\"keywords\\": [
    \\"uuid\\",
    \\"random\\",
    \\"id\\",
    \\"url\\"
  ],
  \\"author\\": \\"Andrey Sitnik <andrey@sitnik.ru>\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"ai/nanoid\\",
  \\"browser\\": {
    \\"./index.js\\": \\"./index.browser.js\\",
    \\"./random.js\\": \\"./random.browser.js\\",
    \\"./async/index.js\\": \\"./async/index.browser.js\\",
    \\"./async/random.js\\": \\"./async/random.browser.js\\"
  },
  \\"sideEffects\\": false
}


export default packages;
"
`;

exports[`loader should not re-read the packages data if nothing changed 2`] = `
"const packages = {};

packages[\\"lodash.isarraylike\\"] = {
  \\"name\\": \\"lodash.isarraylike\\",
  \\"version\\": \\"4.2.0\\",
  \\"description\\": \\"The lodash method \`_.isArrayLike\` exported as a module.\\",
  \\"homepage\\": \\"https://lodash.com/\\",
  \\"icon\\": \\"https://lodash.com/icon.svg\\",
  \\"license\\": \\"MIT\\",
  \\"keywords\\": \\"lodash-modularized, isarraylike\\",
  \\"author\\": \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
  \\"contributors\\": [
    \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
    \\"Blaine Bublitz <blaine.bublitz@gmail.com> (https://github.com/phated)\\",
    \\"Mathias Bynens <mathias@qiwi.be> (https://mathiasbynens.be/)\\"
  ],
  \\"repository\\": \\"lodash/lodash\\",
  \\"scripts\\": { \\"test\\": \\"echo \\\\\\"See https://travis-ci.org/lodash/lodash-cli for testing details.\\\\\\"\\" }
}

packages[\\"nanoid\\"] = {
  \\"name\\": \\"nanoid\\",
  \\"version\\": \\"2.0.3\\",
  \\"description\\": \\"A tiny (141 bytes), secure URL-friendly unique string ID generator\\",
  \\"keywords\\": [
    \\"uuid\\",
    \\"random\\",
    \\"id\\",
    \\"url\\"
  ],
  \\"author\\": \\"Andrey Sitnik <andrey@sitnik.ru>\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"ai/nanoid\\",
  \\"browser\\": {
    \\"./index.js\\": \\"./index.browser.js\\",
    \\"./random.js\\": \\"./random.browser.js\\",
    \\"./async/index.js\\": \\"./async/index.browser.js\\",
    \\"./async/random.js\\": \\"./async/random.browser.js\\"
  },
  \\"sideEffects\\": false
}


export default packages;
"
`;

exports[`loader should update packages data when package.json changes 1`] = `
"const packages = {};

packages[\\"lodash.isarraylike\\"] = {
  \\"name\\": \\"lodash.isarraylike\\",
  \\"version\\": \\"4.2.0\\",
  \\"description\\": \\"The lodash method \`_.isArrayLike\` exported as a module.\\",
  \\"homepage\\": \\"https://lodash.com/\\",
  \\"icon\\": \\"https://lodash.com/icon.svg\\",
  \\"license\\": \\"MIT\\",
  \\"keywords\\": \\"lodash-modularized, isarraylike\\",
  \\"author\\": \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
  \\"contributors\\": [
    \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
    \\"Blaine Bublitz <blaine.bublitz@gmail.com> (https://github.com/phated)\\",
    \\"Mathias Bynens <mathias@qiwi.be> (https://mathiasbynens.be/)\\"
  ],
  \\"repository\\": \\"lodash/lodash\\",
  \\"scripts\\": { \\"test\\": \\"echo \\\\\\"See https://travis-ci.org/lodash/lodash-cli for testing details.\\\\\\"\\" }
}

packages[\\"nanoid\\"] = {
  \\"name\\": \\"nanoid\\",
  \\"version\\": \\"2.0.3\\",
  \\"description\\": \\"A tiny (141 bytes), secure URL-friendly unique string ID generator\\",
  \\"keywords\\": [
    \\"uuid\\",
    \\"random\\",
    \\"id\\",
    \\"url\\"
  ],
  \\"author\\": \\"Andrey Sitnik <andrey@sitnik.ru>\\",
  \\"license\\": \\"MIT\\",
  \\"repository\\": \\"ai/nanoid\\",
  \\"browser\\": {
    \\"./index.js\\": \\"./index.browser.js\\",
    \\"./random.js\\": \\"./random.browser.js\\",
    \\"./async/index.js\\": \\"./async/index.browser.js\\",
    \\"./async/random.js\\": \\"./async/random.browser.js\\"
  },
  \\"sideEffects\\": false
}


export default packages;
"
`;

exports[`loader should update packages data when package.json changes 2`] = `
"const packages = {};

packages[\\"lodash.isarraylike\\"] = {
  \\"name\\": \\"lodash.isarraylike\\",
  \\"version\\": \\"4.2.0\\",
  \\"description\\": \\"The lodash method \`_.isArrayLike\` exported as a module.\\",
  \\"homepage\\": \\"https://lodash.com/\\",
  \\"icon\\": \\"https://lodash.com/icon.svg\\",
  \\"license\\": \\"MIT\\",
  \\"keywords\\": \\"lodash-modularized, isarraylike\\",
  \\"author\\": \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
  \\"contributors\\": [
    \\"John-David Dalton <john.david.dalton@gmail.com> (http://allyoucanleet.com/)\\",
    \\"Blaine Bublitz <blaine.bublitz@gmail.com> (https://github.com/phated)\\",
    \\"Mathias Bynens <mathias@qiwi.be> (https://mathiasbynens.be/)\\"
  ],
  \\"repository\\": \\"lodash/lodash\\",
  \\"scripts\\": { \\"test\\": \\"echo \\\\\\"See https://travis-ci.org/lodash/lodash-cli for testing details.\\\\\\"\\" }
}


export default packages;
"
`;
